src/jupyterlabcontroller/storage/kubernetes/watcher.py

```diff
@@ -0,0 +1,221 @@
+"""Watch a Kubernetes namespace or cluster for events."""
+
+from __future__ import annotations
+
+from collections.abc import AsyncIterator, Awaitable, Callable
+from dataclasses import dataclass
+from datetime import timedelta
+from typing import Any, Generic, Self, TypeVar
+
+from kubernetes_asyncio.client import ApiException
+from kubernetes_asyncio.watch import Watch
+from structlog.stdlib import BoundLogger
+
+from ...exceptions import KubernetesError
+from ...models.domain.kubernetes import WatchEventType
+
+#: Type of Kubernetes object being watched (`dict` for custom objects).
+T = TypeVar("T")
+
+__all__ = [
+    "KubernetesWatcher",
+    "T",
+    "WatchEvent",
+]
+
+
+@dataclass
+class WatchEvent(Generic[T]):
+    """Parsed event from a Kubernetes watch.
+
+    This model is intended only for use within the Kubernetes storage layer.
+    It is interpreted by the callers of the generic watch class inside the
+    per-object Kubernetes storage classes.
+    """
+
+    action: WatchEventType
+    """Action the event represents."""
+
+    object: T
```

Comment:
So...T is just _some_ concrete type, from line 18?

Reply:
Yes, exactly. You can define a bound on the type when you define the `TypeVar`, but in this case it can be any type.